Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift

```diff
@@ -0,0 +1,156 @@
+ Limiting this functionality to `DEBUG` allows us to trade some extra complexity in this file for better _clarity at point of use_ (https://swift.org/documentation/api-design-guidelines/) and guarantee that this doesn't impact release builds.
+ */
+#if DEBUG
+
+extension Resilient {
+
+  @dynamicMemberLookup
+  public struct PropertyLevelErrors {
+    let decodedValue: Value
+
+    /**
+     The error encountered while decoding the value
+     */
+    public let error: Error?
+
+    /**
+     This type defines additional ways decoding errors can be inspected on `Resilient<[T]>` and `Resilient<[T]?>`
+     The properties here are meant to be accessed directly on `PropertyLevelErrors` via `@dynamicMemberLookup`
+     */
+    public struct ArrayErrorDigest<Element> {
+      
+      /**
+       An array of `Result` where elements that failed to decode are represented by `failure` case in the position where they would have occurred had the entire array been decoded successully.
+       If the entire resilient array failed to decode, this will be a single failed `Result` with the top-level error.
+       */
+      public let results: [Result<Element, Error>]
+
+      /**
+       All errors encountered when decoding the resilient array.
+       */
+      public var errors: [Error] {
+        return results.compactMap { result in
+          guard case .failure(let error) = result else {
+            return nil
+          }
+          return error
+        }
+      }
+
+      fileprivate init(error: Error?, elements: [Element]) {
+        switch error {
+        case let error as ArrayDecodingError:
+          results = error.errors(interleavedWith: elements)
+        case let error?:
+          results = [.failure(error)]
+        case nil:
+          results = []
+        }
+      }
+    }
+
+    /**
+     This subscript adds the `errors` and `results` property to `Resilient<[T]>` values using `dynamicMemberLookup`.
+     */
+    public subscript<T, U>(dynamicMember keyPath: KeyPath<ArrayErrorDigest<T>, U>) -> U
+      where
+        Value == [T]
+    {
+      return ArrayErrorDigest(error: error, elements: decodedValue)[keyPath: keyPath]
+    }
+
+    /**
+     This subscript adds the `errors` and `results` property to `Resilient<[T]?>` values using `dynamicMemberLookup`.
+     */
+    public subscript<T, U>(dynamicMember keyPath: KeyPath<ArrayErrorDigest<T>, U>) -> U
+      where
+        Value == [T]?
+    {
+      return ArrayErrorDigest(error: error, elements: decodedValue ?? [])[keyPath: keyPath]
+    }
+  }
+
+}
+
+// MARK: - Array Decoding Error
+
+/**
+ An error which represents any number of errors encountered when decoding a `Resilient` array.
+ */
+struct ArrayDecodingError: Error {
+
+  /**
+   Builds an `ArrayDecodingError`. This type assumes you are decoding elements in order, and care must be taken that the `mutating` members are called in the correct order.
+   */
+  struct Builder {
```

Comment:
What's the reasoning for creating this type instead of initializing the `ArrayDecodingError` directly?

Reply:
You start building an `ArrayDecodingError` before knowing whether it is an error (]or you have successfully decoded the array. The `Builder` encapsulates this logic with its eventual `build()` method returning an optional.

Reply:
What do you think about using an enum to model this type? We can start with a `.none` and then start adding errors through methods that return a new `ArrayDecodingError` with the underlying errors. That will allow us to get rid of the `Builder`. Then, instead of checking for an [optional here](https://github.com/airbnb/ResilientDecoding/pull/11/files#diff-130f54120248da53c6981e26ca46c0feR99) we could check for the errors case.

Reply:
We could even go further and model `ArrayDecodingError` as an enum that holds the errors without the need of a Builder but given [this line](https://github.com/airbnb/ResilientDecoding/pull/11/files#diff-c0df969f14dd120e657b56e5476ee8b0R73) I'm not sure if that's too big of a refactor.

Reply:
Yeah, I do think we need two types though, one that conforms to `Error` and one that doesn't, but vends an `Optional` of the one that does. This was the cleanest way I could achieve this. I'm also not a huge fan of `Builder` types, but at least this one is internal and namespaces to they type its building.